src/components/GeoChart.tsx

```diff
@@ -0,0 +1,90 @@
+// (C) 2019 GoodData Corporation
+import * as React from "react";
+import { VisualizationObject, VisualizationInput } from "@gooddata/typings";
+import omit = require("lodash/omit");
+import noop = require("lodash/noop");
+import get = require("lodash/get");
+
+import { Subtract } from "../typings/subtract";
+import { GeoChart as CoreGeoChart } from "./core/GeoChart";
+import { dataSourceProvider } from "./afm/DataSourceProvider";
+import { ICommonChartProps } from "./core/base/BaseChart";
+import { convertBucketsToAFM, convertBucketsToMdObject } from "../helpers/conversion";
+import { getGeoChartDimensions } from "../helpers/dimensions";
+import { getResultSpec } from "../helpers/resultSpec";
+
+import { COLOR, LOCATION, SEGMENT, SIZE } from "../constants/bucketNames";
+
+export interface IGeoChartBucketProps {
+    location?: VisualizationInput.IAttribute;
+    size?: VisualizationInput.AttributeOrMeasure;
+    color?: VisualizationInput.AttributeOrMeasure;
+    segmentBy?: VisualizationInput.IAttribute;
+    filters?: VisualizationInput.IFilter[];
+    sortBy?: VisualizationInput.ISort[];
+}
+
+export interface IGeoChartProps extends ICommonChartProps, IGeoChartBucketProps {
+    projectId: string;
+    exportTitle?: string;
+}
+
+export const getBuckets = (props: IGeoChartProps): VisualizationObject.IBucket[] => {
+    const { color, location, segmentBy, size, config } = props;
+    const bucket = [
+        {
+            localIdentifier: SIZE,
+            items: size ? [size] : [],
+        },
+        {
+            localIdentifier: COLOR,
+            items: color ? [color] : [],
+        },
+        {
+            localIdentifier: LOCATION,
+            items: location ? [location] : [],
+        },
+        {
+            localIdentifier: SEGMENT,
+            items: segmentBy ? [segmentBy] : [],
+        },
+    ];
+    const tooltipText = get(config, "tooltipText");
+    if (tooltipText) {
+        bucket.push({
+            localIdentifier: "tooltipText",
+            items: [tooltipText] || [],
+        });
+    }
+    return bucket;
+};
+
+// noop is never called because resultSpec is always provided
+const DataSourceProvider = dataSourceProvider(CoreGeoChart, noop as any, "GeoChart");
+
+type IGeoChartNonBucketProps = Subtract<IGeoChartProps, IGeoChartBucketProps>;
+
+export class GeoChart extends React.Component<IGeoChartProps> {
```

Comment:
please convert this to functional component, here we don't need constructor and lifecycle method